src/stdlib_linalg_determinant.fypp

```diff
@@ -0,0 +1,234 @@
+                   case (1:)
+                       err0 = linalg_state_type(this,LINALG_ERROR,'singular matrix')
+                   case default
+                       err0 = linalg_state_type(this,LINALG_INTERNAL_ERROR,'catastrophic error')
+                end select
+
+                deallocate(amat)
+
+         end select
+
+         ! Process output and return
+         1 call linalg_error_handling(err0)
+
+     end function stdlib_linalg_pure_${rt[0]}$${rk}$determinant
+     
+     module function stdlib_linalg_${rt[0]}$${rk}$determinant(a,overwrite_a,err) result(det)
+     !!### Summary
+     !! Compute determinant of a square matrix (with error control).
+     !!
+     !!### Description
+     !!
+     !! This function computes the determinant of a square matrix with error control.
+     !!
+     !! param: a Input matrix of size [m,n].
+     !! param: overwrite_a [optional] Flag indicating if the input matrix can be overwritten.
+     !! param: err State return flag. 
+     !! return: det Matrix determinant.
+     !!
+     !!### Example
+     !!
+     !!```fortran
+     !!
+     !! ${rt}$ :: matrix(3,3)
+     !! ${rt}$ :: determinant
+     !! matrix = reshape([1, 2, 3, 4, 5, 6, 7, 8, 9], [3, 3])
+     !! determinant = det(matrix, err=err)         
+     !!
+     !!```        
+     !     
+         !> Input matrix a[m,n]
+         ${rt}$, intent(inout), target :: a(:,:)
+         !> [optional] Can A data be overwritten and destroyed?
+         logical(lk), optional, intent(in) :: overwrite_a
+         !> State return flag. 
+         type(linalg_state_type), intent(out) :: err
+         !> Matrix determinant
+         ${rt}$ :: det
+
+         !! Local variables
+         type(linalg_state_type) :: err0
+         integer(ilp) :: m,n,info,perm,k
+         integer(ilp), allocatable :: ipiv(:)
+         logical(lk) :: copy_a
+         ${rt}$, pointer :: amat(:,:)
+
+         ! Matrix determinant size
+         m = size(a,1,kind=ilp)
+         n = size(a,2,kind=ilp)
+
+         if (m/=n .or. .not.min(m,n)>=0) then
+            err0 = linalg_state_type(this,LINALG_VALUE_ERROR,'invalid or non-square matrix: a=[',m,',',n,']')
+            det  = 0.0_${rk}$
+            goto 1
+         end if
+
+         ! Can A be overwritten? By default, do not overwrite
+         if (present(overwrite_a)) then
+            copy_a = .not.overwrite_a
+         else
+            copy_a = .true._lk
+         endif
+
+         select case (m)
+            case (0)
+                
+                ! Empty array has determinant 1 because math
+                det = 1.0_${rk}$
+
+            case (1)
+                
+                ! Scalar input
+                det = a(1,1)
+
+            case default
+
+                ! Find determinant from LU decomposition
+
+                ! Initialize a matrix temporary
+                if (copy_a) then
+                   allocate(amat(m,n),source=a)
```

Comment:
I agree with that. I always add the size in sourced allocation because [old gfortran](https://godbolt.org/z/csM3PPTKa) versions do not supported, but because it is <=5, we should probably not worry about that.